sql/2019/04_Media/04_08.sql

```diff
@@ -1,73 +1,17 @@
-I extract the srcset tag AND the sizes tag FROM each img, AND determine how many sources are present 
-(I am COUNTing commas AND then adding one to get the total, so commas in the url will screw with the numbers)
-the ratio is simply sources/sizes.
-
-I then break it into percentiles to see the breakdown of source images AND sizes AND ratios.  
-I also supply the COUNT of scrset AND size tags in the dataset.
-
-04_08
-I added a query for the src tag... AND then feed theat up through the same queries for 04_06. 
-in the 1k mobile sample set 2918 have srcset, AND 2932 have a fallback src
-
-04_13
-added in alt tag query.  get COUNT AND LENGTH - note that LENGTH 6 = alt="" (empty)
-
-*/
-
-
-SELECT client, APPROX_QUANTILES(srcsetCOUNT,10) AS srcsetperc, 
-	APPROX_QUANTILES(sizesCOUNT,10) AS sizesperc, 
-	APPROX_QUANTILES(ratio,10) AS ratioperc,
-	APPROX_QUANTILES(srcCOUNT,10) AS srcperc,
-	APPROX_QUANTILES(altLENGTH,10) AS altTextLENGTHPercentiles,
-  	COUNT(srcset) AS srcsetCOUNT, 
- 	COUNT(sizes) AS  sizeCOUNT, 
- 	COUNT(srcCOUNT) AS srcCOUNT, 
- 	COUNT(altCOUNT) AS altCOUNT
-
-FROM(
-
-SELECT 
-     client, 
-     #this COUNTs the number of commas, AND adds one - to get the number of srcsets
-     LENGTH(srcset) - LENGTH(REGEXP_REPLACE(srcset, ',', '')) +1 srcsetCOUNT,
-     #this COUNTs the number of commas, AND adds one - to get the number of sizes
-     LENGTH(sizes) - LENGTH(REGEXP_REPLACE(sizes, ',', '')) +1 sizesCOUNT, 
-     #ratio of srcset to sizes
-     (LENGTH(srcset) - LENGTH(REGEXP_REPLACE(srcset, ',', '')) +1)/(LENGTH(sizes) - LENGTH(REGEXP_REPLACE(sizes, ',', '')) +1) ratio
-     ,
-     LENGTH(src)/LENGTH(src) srcCOUNT,
-     LENGTH(alt)/LENGTH(alt) altCOUNT, LENGTH(alt) altLENGTH,
-     #double check that its working
-     srcset, sizes, alt
-FROM(
-
-SELECT client, REGEXP_EXTRACT(flat_img,r'srcset=".*?"') srcset, REGEXP_EXTRACT(flat_img,r'sizes=".*?"') sizes,
-REGEXP_EXTRACT(flat_img,r'src=".*?"') src, REGEXP_EXTRACT(flat_img,r'alt=".*?"') alt
-
-#here I am extracting the srcset=" *** " AND sizes " *** " for analysis   
-FROM(
-SELECT client, url, flat_img 
-#this CROSS JOIN flattens the array of image tags into strings.
-FROM(
-
-#this query grabs all img tags with srcset inside. it also grabs the same for the picture tag
-SELECT url, REGEXP_EXTRACT_ALL(body,r'(<img.*?srcset.*?>)') img,REGEXP_EXTRACT_ALL(body,r'(<picture.*?srcset.*?/picture>)') picture,  _TABLE_SUFFIX AS client
-FROM `response_bodies.2019_07_01_*`  srcsets
-WHERE body LIKE "%srcset%sizes%")
-CROSS JOIN UNNEST(img) flat_img
-)
-)
-)GROUP BY client
\ No newline at end of file
+# 04_08: Pages with <picture><img></picture>
+SELECT
+  client,
+  COUNTIF(has_picture_img) AS has_picture_img,
+  COUNT(0) AS total,
+  ROUND(COUNTIF(has_source_sizes) * 100 / COUNT(0), 2) AS pct
+FROM (
+  SELECT
+    client,
+    REGEXP_CONTAINS(body, r'(?si)<picture.*?<img.*?/picture>') AS has_picture_img
```

Comment:
While perhaps rare, this will match the following. Are we ok with that?

```html
<picture>
  <source srcset="/static/source-image.jpg">
</picture>

<img src="/static/non-source-image.jpg" />

<picture>
  <source srcset="/static/source-image.jpg">
</picture>
```

Reply:
Good catch. Yeah parsing HTML with regular expressions is extremely suboptimal and in an ideal world these would have been implemented reliably as custom metrics. If and when we reuse these metrics for future editions of the Almanac we should definitely switch to that approach. For now, I can accept the side-effects.